In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [2]:
train_data = pd.read_csv('C:/Users/Kiran Shankar Bhat/Documents/Mercedes-Benz/train.csv')

In [3]:
test_data = pd.read_csv('C:/Users/Kiran Shankar Bhat/Documents/Mercedes-Benz/test.csv')

In [4]:
train_data.shape

(4209, 378)

In [5]:
train_data.head(2)

ID       y X0 X1  X2 X3 X4 X5 X6 X8  ...  X375  X376  X377  X378  X379  \
0   0  130.81  k  v  at  a  d  u  j  o  ...     0     0     1     0     0   
1   6   88.53  k  t  av  e  d  y  l  o  ...     1     0     0     0     0   

   X380  X382  X383  X384  X385  
0     0     0     0     0     0  
1     0     0     0     0     0  

[2 rows x 378 columns]

In [87]:
test_data.head(2)

ID  X0 X1  X2 X3 X4 X5 X6 X8  X10  ...  X375  X376  X377  X378  X379  X380  \
0   1  az  v   n  f  d  t  a  w    0  ...     0     0     0     1     0     0   
1   2   t  b  ai  a  d  b  g  y    0  ...     0     0     1     0     0     0   

   X382  X383  X384  X385  
0     0     0     0     0  
1     0     0     0     0  

[2 rows x 372 columns]

In [6]:
test_data.shape

(4209, 377)

In [7]:
type(train_data.var())

pandas.core.series.Series

In [8]:
train_data.var()

ID      5.941936e+06
y       1.607667e+02
X10     1.313092e-02
X11     0.000000e+00
X12     6.945713e-02
            ...     
X380    8.014579e-03
X382    7.546747e-03
X383    1.660732e-03
X384    4.750593e-04
X385    1.423823e-03
Length: 370, dtype: float64

In [9]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4209 entries, 0 to 4208
Columns: 378 entries, ID to X385
dtypes: float64(1), int64(369), object(8)
memory usage: 12.1+ MB


In [10]:
train_data.dtypes

ID        int64
y       float64
X0       object
X1       object
X2       object
         ...   
X380      int64
X382      int64
X383      int64
X384      int64
X385      int64
Length: 378, dtype: object

In [11]:
train_data.columns

Index(['ID', 'y', 'X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8',
       ...
       'X375', 'X376', 'X377', 'X378', 'X379', 'X380', 'X382', 'X383', 'X384',
       'X385'],
      dtype='object', length=378)

# Checking for variance value

In [12]:
train_var_data = train_data.copy()

In [13]:
train_var_data.head(3)

ID       y  X0 X1  X2 X3 X4 X5 X6 X8  ...  X375  X376  X377  X378  X379  \
0   0  130.81   k  v  at  a  d  u  j  o  ...     0     0     1     0     0   
1   6   88.53   k  t  av  e  d  y  l  o  ...     1     0     0     0     0   
2   7   76.26  az  w   n  c  d  x  j  x  ...     0     0     0     0     0   

   X380  X382  X383  X384  X385  
0     0     0     0     0     0  
1     0     0     0     0     0  
2     0     1     0     0     0  

[3 rows x 378 columns]

In [14]:
train_var_data.drop(['X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8'],axis=1,inplace=True)

In [15]:
train_var_data.columns[train_var_data.var()==0]

Index(['X11', 'X93', 'X107', 'X233', 'X235', 'X268', 'X289', 'X290', 'X293',
       'X297', 'X330', 'X347'],
      dtype='object')

we got 12 columns which are having the variance value 0 

In [16]:
train_var_data.drop(['X11', 'X93', 'X107', 'X233', 'X235', 'X268', 'X289', 'X290', 'X293',
       'X297', 'X330', 'X347'],axis=1,inplace=True)

In [17]:
train_var_data.shape

(4209, 358)

For test data

In [18]:
test_data.head(3)

ID  X0 X1  X2 X3 X4 X5 X6 X8  X10  ...  X375  X376  X377  X378  X379  X380  \
0   1  az  v   n  f  d  t  a  w    0  ...     0     0     0     1     0     0   
1   2   t  b  ai  a  d  b  g  y    0  ...     0     0     1     0     0     0   
2   3  az  v  as  f  d  a  j  j    0  ...     0     0     0     1     0     0   

   X382  X383  X384  X385  
0     0     0     0     0  
1     0     0     0     0  
2     0     0     0     0  

[3 rows x 377 columns]

In [19]:
test_data.dtypes

ID       int64
X0      object
X1      object
X2      object
X3      object
         ...  
X380     int64
X382     int64
X383     int64
X384     int64
X385     int64
Length: 377, dtype: object

In [20]:
test_data.columns

Index(['ID', 'X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8', 'X10',
       ...
       'X375', 'X376', 'X377', 'X378', 'X379', 'X380', 'X382', 'X383', 'X384',
       'X385'],
      dtype='object', length=377)

In [21]:
test_var_data = test_data.copy()

In [22]:
test_var_data.drop(['X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8'],axis=1,inplace=True)

In [23]:
test_var_data.columns[test_var_data.var() == 0]

Index(['X257', 'X258', 'X295', 'X296', 'X369'], dtype='object')

We got 5 columns which are having varence value = 0

In [24]:
test_var_data.drop(['X257', 'X258', 'X295', 'X296', 'X369'],axis=1,inplace=True)

In [25]:
test_var_data.head()

ID  X10  X11  X12  X13  X14  X15  X16  X17  X18  ...  X375  X376  X377  \
0   1    0    0    0    0    0    0    0    0    0  ...     0     0     0   
1   2    0    0    0    0    0    0    0    0    0  ...     0     0     1   
2   3    0    0    0    0    1    0    0    0    0  ...     0     0     0   
3   4    0    0    0    0    0    0    0    0    0  ...     0     0     0   
4   5    0    0    0    0    1    0    0    0    0  ...     1     0     0   

   X378  X379  X380  X382  X383  X384  X385  
0     1     0     0     0     0     0     0  
1     0     0     0     0     0     0     0  
2     1     0     0     0     0     0     0  
3     1     0     0     0     0     0     0  
4     0     0     0     0     0     0     0  

[5 rows x 364 columns]

# Null and Unique values

Chicking for null and unique values in test and train datasets

In [26]:
train_data.head()

ID       y  X0 X1  X2 X3 X4 X5 X6 X8  ...  X375  X376  X377  X378  X379  \
0   0  130.81   k  v  at  a  d  u  j  o  ...     0     0     1     0     0   
1   6   88.53   k  t  av  e  d  y  l  o  ...     1     0     0     0     0   
2   7   76.26  az  w   n  c  d  x  j  x  ...     0     0     0     0     0   
3   9   80.62  az  t   n  f  d  x  l  e  ...     0     0     0     0     0   
4  13   78.02  az  v   n  f  d  h  d  n  ...     0     0     0     0     0   

   X380  X382  X383  X384  X385  
0     0     0     0     0     0  
1     0     0     0     0     0  
2     0     1     0     0     0  
3     0     0     0     0     0  
4     0     0     0     0     0  

[5 rows x 378 columns]

In [27]:
train_data.columns[train_data.isnull().any()]

Index([], dtype='object')

In [28]:
test_data.columns[test_data.isnull().any()]

Index([], dtype='object')

We don't have any  Null values in testing and training data set

Checking for unique values in testing and training data

In [29]:
train_data.nunique()

ID      4209
y       2545
X0        47
X1        27
X2        44
        ... 
X380       2
X382       2
X383       2
X384       2
X385       2
Length: 378, dtype: int64

In training dataset other than columns 'ID' and 'y' all othere columns have only 2 unique values i.e 0 & 1. 
y has 2545 unique values

In [30]:
train_data['y'].unique()

array([130.81,  88.53,  76.26, ...,  85.71, 108.77,  87.48])

In [31]:
test_data.nunique()

ID      4209
X0        49
X1        27
X2        45
X3         7
        ... 
X380       2
X382       2
X383       2
X384       2
X385       2
Length: 377, dtype: int64

In testing data other than 'ID' all other columns has 2 unique values i.e 0 & 1 

# Applying Label encoder for catorigical columns

Removing the columns that are having the variance value = 0 From the main train and testing dataset

In [32]:
train_data.drop(['X11', 'X93', 'X107', 'X233', 'X235', 'X268', 'X289', 'X290', 'X293',
       'X297', 'X330', 'X347'],axis=1,inplace=True)

In [33]:
test_data.drop(['X257', 'X258', 'X295', 'X296', 'X369'],axis=1,inplace=True)

In [34]:
train_obj = train_data.select_dtypes(include=['object']).copy()
train_obj.head()

X0 X1  X2 X3 X4 X5 X6 X8
0   k  v  at  a  d  u  j  o
1   k  t  av  e  d  y  l  o
2  az  w   n  c  d  x  j  x
3  az  t   n  f  d  x  l  e
4  az  v   n  f  d  h  d  n

In [35]:
train_obj.columns

Index(['X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8'], dtype='object')

In [36]:
train_obj.nunique()

X0    47
X1    27
X2    44
X3     7
X4     4
X5    29
X6    12
X8    25
dtype: int64

In [37]:
test_obj = test_data.select_dtypes(include=['object']).copy()
test_obj.head()

X0 X1  X2 X3 X4 X5 X6 X8
0  az  v   n  f  d  t  a  w
1   t  b  ai  a  d  b  g  y
2  az  v  as  f  d  a  j  j
3  az  l   n  f  d  z  l  n
4   w  s  as  c  d  y  i  m

In [38]:
test_obj.columns

Index(['X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8'], dtype='object')

In [39]:
test_obj.nunique()

X0    49
X1    27
X2    45
X3     7
X4     4
X5    32
X6    12
X8    25
dtype: int64

The columns X0,X1.X2,X3,X4,X5,X6,X8 are having catogerical data in both test and training dataset

In [40]:
from sklearn.preprocessing import LabelEncoder

In [41]:
lable = LabelEncoder()

In [42]:
train_obj['l_X0'] = lable.fit_transform(train_obj['X0'])
train_obj['l_X1'] = lable.fit_transform(train_obj['X1'])
train_obj['l_X2'] = lable.fit_transform(train_obj['X2'])
train_obj['l_X3'] = lable.fit_transform(train_obj['X3'])
train_obj['l_X4'] = lable.fit_transform(train_obj['X4'])
train_obj['l_X5'] = lable.fit_transform(train_obj['X5'])
train_obj['l_X6'] = lable.fit_transform(train_obj['X6'])
train_obj['l_X8'] = lable.fit_transform(train_obj['X8'])

In [43]:
train_obj.head()

X0 X1  X2 X3 X4 X5 X6 X8  l_X0  l_X1  l_X2  l_X3  l_X4  l_X5  l_X6  l_X8
0   k  v  at  a  d  u  j  o    32    23    17     0     3    24     9    14
1   k  t  av  e  d  y  l  o    32    21    19     4     3    28    11    14
2  az  w   n  c  d  x  j  x    20    24    34     2     3    27     9    23
3  az  t   n  f  d  x  l  e    20    21    34     5     3    27    11     4
4  az  v   n  f  d  h  d  n    20    23    34     5     3    12     3    13

In [44]:
train_le_data = train_obj.drop(['X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8'],axis=1)

In [45]:
train_le_data.head()

l_X0  l_X1  l_X2  l_X3  l_X4  l_X5  l_X6  l_X8
0    32    23    17     0     3    24     9    14
1    32    21    19     4     3    28    11    14
2    20    24    34     2     3    27     9    23
3    20    21    34     5     3    27    11     4
4    20    23    34     5     3    12     3    13

For test data set

In [46]:
test_obj['l_X0'] = lable.fit_transform(test_obj['X0'])
test_obj['l_X1'] = lable.fit_transform(test_obj['X1'])
test_obj['l_X2'] = lable.fit_transform(test_obj['X2'])
test_obj['l_X3'] = lable.fit_transform(test_obj['X3'])
test_obj['l_X4'] = lable.fit_transform(test_obj['X4'])
test_obj['l_X5'] = lable.fit_transform(test_obj['X5'])
test_obj['l_X6'] = lable.fit_transform(test_obj['X6'])
test_obj['l_X8'] = lable.fit_transform(test_obj['X8'])

In [47]:
test_obj.head()

X0 X1  X2 X3 X4 X5 X6 X8  l_X0  l_X1  l_X2  l_X3  l_X4  l_X5  l_X6  l_X8
0  az  v   n  f  d  t  a  w    21    23    34     5     3    26     0    22
1   t  b  ai  a  d  b  g  y    42     3     8     0     3     9     6    24
2  az  v  as  f  d  a  j  j    21    23    17     5     3     0     9     9
3  az  l   n  f  d  z  l  n    21    13    34     5     3    31    11    13
4   w  s  as  c  d  y  i  m    45    20    17     2     3    30     8    12

In [48]:
test_le_data = test_obj.drop(['X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8'],axis=1)

In [49]:
test_le_data.head()

l_X0  l_X1  l_X2  l_X3  l_X4  l_X5  l_X6  l_X8
0    21    23    34     5     3    26     0    22
1    42     3     8     0     3     9     6    24
2    21    23    17     5     3     0     9     9
3    21    13    34     5     3    31    11    13
4    45    20    17     2     3    30     8    12

# Scaling/standardisation the encoded columns

In [50]:
from sklearn.preprocessing import StandardScaler

In [51]:
scale = StandardScaler()

In [52]:
scale.fit(train_le_data)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [53]:
scaled_train = scale.transform(train_le_data)

In [54]:
scaled_train

array([[ 0.16301209,  1.39348787, -0.02812155, ...,  1.29211729,
         0.75178725,  0.33944483],
       [ 0.16301209,  1.15902093,  0.15538793, ...,  1.77697445,
         1.43751106,  0.33944483],
       [-0.71055977,  1.51072134,  1.53170902, ...,  1.65576016,
         0.75178725,  1.61838949],
       ...,
       [-1.58413164,  1.39348787,  1.89872798, ..., -1.61702573,
        -0.27679847, -1.08160479],
       [-1.51133398,  0.924554  ,  0.70591637, ..., -1.61702573,
         1.43751106,  1.1920746 ],
       [ 1.18217927,  0.924554  , -1.31268791, ..., -1.61702573,
        -0.27679847,  1.47628453]])

In [55]:
train_scale_data = pd.DataFrame(scaled_train,columns=['ls_X0', 'ls_X1', 'ls_X2', 'ls_X3', 'ls_X4', 'ls_X5', 'ls_X6', 'ls_X8'])

In [56]:
train_scale_data.head()

ls_X0     ls_X1     ls_X2     ls_X3     ls_X4     ls_X5     ls_X6  \
0  0.163012  1.393488 -0.028122 -1.678270  0.028938  1.292117  0.751787   
1  0.163012  1.159021  0.155388  0.620969  0.028938  1.776974  1.437511   
2 -0.710560  1.510721  1.531709 -0.528650  0.028938  1.655760  0.751787   
3 -0.710560  1.159021  1.531709  1.195779  0.028938  1.655760  1.437511   
4 -0.710560  1.393488  1.531709  1.195779  0.028938 -0.162454 -1.305384   

      ls_X8  
0  0.339445  
1  0.339445  
2  1.618389  
3 -1.081605  
4  0.197340

Here ,ls_X0...lsX8 is scaled train features

In [57]:
scale.fit(test_le_data)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [58]:
scaled_test = scale.transform(test_le_data)

In [59]:
scaled_test

array([[-0.62521149,  1.39576032,  1.58606761, ...,  1.26665157,
        -2.38888808,  1.48839606],
       [ 0.75460919, -0.94519929, -0.95644521, ..., -0.69501069,
        -0.2966019 ,  1.77347725],
       [-0.62521149,  1.39576032, -0.07634462, ..., -1.73353777,
         0.74954118, -0.36463164],
       ...,
       [ 1.08313793,  1.39576032, -0.07634462, ..., -1.61814587,
        -1.34274499,  1.48839606],
       [-1.54509194,  1.39576032, -0.07634462, ..., -1.61814587,
        -1.69145935,  0.63315251],
       [ 0.75460919, -1.17929525, -0.95644521, ..., -1.61814587,
        -0.2966019 ,  0.7756931 ]])

In [60]:
test_scale_data = pd.DataFrame(scaled_test,columns=['ls_X0', 'ls_X1', 'ls_X2', 'ls_X3', 'ls_X4', 'ls_X5', 'ls_X6', 'ls_X8'])

In [61]:
test_scale_data.head()

ls_X0     ls_X1     ls_X2     ls_X3     ls_X4     ls_X5     ls_X6  \
0 -0.625211  1.395760  1.586068  1.163082  0.036299  1.266652 -2.388888   
1  0.754609 -0.945199 -0.956445 -1.651020  0.036299 -0.695011 -0.296602   
2 -0.625211  1.395760 -0.076345  1.163082  0.036299 -1.733538  0.749541   
3 -0.625211  0.225281  1.586068  1.163082  0.036299  1.843611  1.446970   
4  0.951726  1.044616 -0.076345 -0.525379  0.036299  1.728219  0.400827   

      ls_X8  
0  1.488396  
1  1.773477  
2 -0.364632  
3  0.205531  
4  0.062990

Here, ls_X0...lsX8 is scaled test features

# Concating the scaled train and test datasets to main train & test dataset

In [62]:
train_data.head()

ID       y  X0 X1  X2 X3 X4 X5 X6 X8  ...  X375  X376  X377  X378  X379  \
0   0  130.81   k  v  at  a  d  u  j  o  ...     0     0     1     0     0   
1   6   88.53   k  t  av  e  d  y  l  o  ...     1     0     0     0     0   
2   7   76.26  az  w   n  c  d  x  j  x  ...     0     0     0     0     0   
3   9   80.62  az  t   n  f  d  x  l  e  ...     0     0     0     0     0   
4  13   78.02  az  v   n  f  d  h  d  n  ...     0     0     0     0     0   

   X380  X382  X383  X384  X385  
0     0     0     0     0     0  
1     0     0     0     0     0  
2     0     1     0     0     0  
3     0     0     0     0     0  
4     0     0     0     0     0  

[5 rows x 366 columns]

In [63]:
drop_train = train_data.copy()

In [64]:
drop_train.drop(['X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8'],axis=1,inplace=True)

In [65]:
drop_test = test_data.copy()

In [66]:
drop_test.drop(['X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8'],axis=1,inplace=True)

In [67]:
s_train = pd.concat([drop_train ,train_scale_data],axis=1)

In [68]:
s_test = pd.concat([drop_test,test_scale_data],axis=1)

In [69]:
s_train.shape

(4209, 366)

# Demension reduction

In [70]:
from sklearn.decomposition import PCA

In [71]:
train_pca = PCA(n_components = 4)

In [72]:
test_pca = PCA(n_components= 4 )

In [73]:
train_pca.fit(s_train)

PCA(copy=True, iterated_power='auto', n_components=4, random_state=None,
    svd_solver='auto', tol=0.0, whiten=False)

In [74]:
final_train_data = train_pca.transform(s_train)

In [75]:
final_train_data.shape

(4209, 4)

In [76]:
test_pca.fit(s_test)

PCA(copy=True, iterated_power='auto', n_components=4, random_state=None,
    svd_solver='auto', tol=0.0, whiten=False)

In [77]:
final_test_data = test_pca.transform(s_test)

In [78]:
final_test_data.shape

(4209, 4)

In [79]:
from sklearn.model_selection import train_test_split

In [80]:
x_train,x_test,y_train,y_test = train_test_split(final_train_data,final_test_data,test_size=0.25)

In [81]:
from xgboost import XGBRegressor

In [82]:
xg_boost = XGBRegressor()

In [83]:
xg_boost.fit(x_train,y_train)

[11:23:19] WARNING: src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.


XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=1, gamma=0,
             importance_type='gain', learning_rate=0.1, max_delta_step=0,
             max_depth=3, min_child_weight=1, missing=None, n_estimators=100,
             n_jobs=1, nthread=None, objective='reg:linear', random_state=0,
             reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
             silent=None, subsample=1, verbosity=1)

In [84]:
y_pre = xg_boost.predict(x_test)

In [85]:
y_pre

array([257.47266 , -37.574688, 157.801   , ...,  60.12029 , 163.53941 ,
        20.870144], dtype=float32)

In [86]:
xg_boost.score(x_test,y_pre)

1.0